### PR TITLE
Add org.godotengine.Godot3 (Godot 3 Flatpak to be maintained in tandem with current Godot 4 Flatpak)

### PR DESCRIPTION
**Pull request by Zishan-Rahman**
For now, I have the Godot 3 Flatpak on here as a submodule of my own fork of the existing Godot Flatpak, so that no prior commit history is erased. In practice, would that be a problem and is there a better way to go about this (while preserving prior commit history, of course)? Also, it goes without saying, but I hope I didn't botch this somehow!

Other than that, this PR addresses #3960 and once we've sorted this out, we should be able to close that issue afterwards. If you have any other feedback to give to this newbie, please do so as I'm always willing to learn!

### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** https://github.com/flathub/org.godotengine.Godot/issues/115#issuecomment-1470915933
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by hfiguiere** [comment]
this won't work, there is no top level manifest properly named.

**Comment by Zishan-Rahman** [comment]
> this won't work, there is no top level manifest properly named.

@hfiguiere Hmmmm, you're right. So how do I move the stuff in my Godot Flatpak repo to my Flathub fork, preferably without erasing the commit history? Is that even possible?

**Comment by hfiguiere** [comment]
My take is this https://github.com/flathub/flathub/issues/3960#issuecomment-1456612751

**Comment by Zishan-Rahman** [comment]
> My take is this [#3960 (comment)](https://github.com/flathub/flathub/issues/3960#issuecomment-1456612751)

@hfiguiere This is a pretty good idea, but the thing is that the **current** Godot Flatpak is _already_ at 4.0, so to implement it we'll have to rename the current Flatpak and all references in it to `org.godotengine.Godot4`, and we'll need the approval of some of the other maintainers for that, including @Calinou (who suggested the Godot3 Flatpak I'm trying to do in this PR).

> So how do I move the stuff in my Godot Flatpak repo to my Flathub fork, preferably without erasing the commit history? Is that even possible? And is there something else I'm missing in this equation?

I wonder what @Calinou's take on _this_ would be. If anyone else can chime in, please do.

**Comment by hfiguiere** [comment]
otherwise all flatpak repo are likely to be a fork of this anyway, so you should be able to just do a pull request after playing with the remotes.

**Comment by hfiguiere** [comment]
the filenames have to match the app-id.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# Flatpak for Godot Engine 3 (LTS)
```
[the remark this answers is not in the document]
Comment (Zishan-Rahman) [reply README.md:1]: > this stuff will never be seen by users...

I've taken out the (LTS) bit

**Comment by Zishan-Rahman** [comment]
> the filenames have to match the app-id.

Ooh, how did I miss that? :man_facepalming: 

Fixed now :+1:

File: org.godotengine.Godot3.appdata.xml [hunk 2]
```diff
@@ -0,0 +1,66 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>org.godotengine.Godot.desktop</id>
```
Comment (hfiguiere) [inline org.godotengine.Godot3.appdata.xml:3]:
```suggestion
  <id>org.godotengine.Godot</id>
```

File: org.godotengine.Godot3.yaml [hunk 3]
```diff
@@ -0,0 +1,100 @@
+app-id: org.godotengine.Godot3
+runtime: org.freedesktop.Platform
+runtime-version: '22.08'
+default-branch: stable
```
Comment (hfiguiere) [inline org.godotengine.Godot3.yaml:4]:
```suggestion
```

File: org.godotengine.Godot3.yaml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      - desktop-file-edit --set-icon=$FLATPAK_ID misc/dist/linux/$FLATPAK_ID.desktop
+      - install -Dm644 misc/dist/linux/$FLATPAK_ID.desktop /app/share/applications/$FLATPAK_ID.desktop
+      - install -Dm644 misc/dist/linux/$FLATPAK_ID.xml /app/share/mime/packages/$FLATPAK_ID.xml
+      - install -Dm644 $FLATPAK_ID.appdata.xml /app/share/appdata/$FLATPAK_ID.appdata.xml
+      - install -Dm644 icon.svg /app/share/icons/hicolor/scalable/apps/$FLATPAK_ID.svg
+      - >
+        for size in {32,64,128,256}; do
+          rsvg-convert icon.svg -w "$size" -h "$size" -a -f png -o "$size.png";
+          install -Dm644 "$size.png" "/app/share/icons/hicolor/${size}x${size}/apps/$FLATPAK_ID.png";
+        done
```
Comment (hfiguiere) [inline org.godotengine.Godot3.yaml:100]: you don't even need to do that. All you need is the SVG.

Comment (Zishan-Rahman) [reply org.godotengine.Godot3.yaml:100]: > you don't even need to do that. All you need is the SVG.

Ah. That's in the original Flatpak as well. I assume you want me to just take out the lines without replacing them, so I'll do that. :+1:

**Comment by Zishan-Rahman** [comment]
Well, that's strange. [This build's still in the queue](https://buildbot.flathub.org/#/buildrequests/110788) and it's already been over half an hour. Meanwhile, other Flatpaks are being added to the queue and then built instantaneously. Is it due to the way [the Flathub CI/CD](https://buildbot.flathub.org/#/builders/6) prioritises builds? Or is there something else I'm missing?

In the meantime, should I implement the suggested changes in the code reviews and hope that a new build get picked up quicker? Or should I just wait until our ongoing build request gets processed to push said changes?

**Comment by Zishan-Rahman** [comment]
bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
Nevermind, the build did itself somehow. I guess _all_ these builds get tended to eventually by the CI/CD. Probably because of the "Users" column I got the impression that we could somehow get the builds to start ourselves.



> Build [28829 failed](https://buildbot.flathub.org/#/builders/6/builds/28829)

I know why this build failed and have just pushed a fix.

**Comment by Zishan-Rahman** [comment]
bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
The ongoing build is about to fail because of the way I applied the patches. Maybe these new changes will help?

**Comment by Zishan-Rahman** [comment]
I'm still waiting on [this test build](https://buildbot.flathub.org/#/buildrequests/110964) to start.

**Comment by Zishan-Rahman** [comment]
bot, build org.godotengine.Godot3

File: org.godotengine.Godot3.appdata.xml [hunk 5]
```diff
@@ -0,0 +1,66 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>org.godotengine.Godot</id>
```
Comment (hfiguiere) [inline org.godotengine.Godot3.appdata.xml:3]: this does not match

File: org.godotengine.Godot3.yaml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+      - type: script
+        dest-filename: godot.sh
+        commands:
+          - export APPDATA="$XDG_DATA_HOME"
+          - export PATH="/app/jre/bin:$PATH"
+          - /app/bin/godot-bin "$@"
+
+      - type: file
+        path: org.godotengine.Godot3.appdata.xml
```
Comment (hfiguiere) [inline org.godotengine.Godot3.yaml:83]: THIS file is never installed.

File: org.godotengine.Godot3.yaml [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    build-commands:
+      - python3 /app/bin/scons $SCONS_FLAGS $SCONS_FLAGS_EXTRA tools=yes target=release_debug -j "$FLATPAK_BUILDER_N_JOBS"
+      - install -D -m755 bin/godot.x11.opt.tools.* /app/bin/godot-bin
+      - install -D -m755 godot.sh /app/bin/godot
+      - mv misc/dist/linux/org.godotengine.Godot.desktop misc/dist/linux/$FLATPAK_ID.desktop
+      - mv misc/dist/linux/org.godotengine.Godot.xml misc/dist/linux/$FLATPAK_ID.xml
```
Comment (hfiguiere) [inline org.godotengine.Godot3.yaml:90]: you'll need to change the app-id in the appdata file.

File: org.godotengine.Godot3.yaml (hunk 6 above)
Comment (hfiguiere) [reply org.godotengine.Godot3.yaml:83]: oh it's actually duplicate

File: org.godotengine.Godot3.yaml (hunk 7 above)
Comment (hfiguiere) [reply org.godotengine.Godot3.yaml:90]: also there are duplicates. line 89 and 92. lin 90 and 94.

Comment (Zishan-Rahman) [reply org.godotengine.Godot3.yaml:90]: > you'll need to change the app-id in the appdata file.

You mean to tell me _that_ was why it wasn't working? :man_facepalming:

Comment (hfiguiere) [inline org.godotengine.Godot3.yaml:90]:
```suggestion
```

No point in renaming.

File: org.godotengine.Godot3.yaml [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+      - desktop-file-edit --set-icon=org.godotengine.Godot misc/dist/linux/$FLATPAK_ID.desktop
+      - install -Dm644 misc/dist/linux/$FLATPAK_ID.desktop /app/share/applications/$FLATPAK_ID.desktop
+      - install -Dm644 misc/dist/linux/$FLATPAK_ID.xml /app/share/mime/packages/$FLATPAK_ID.xml
```
Comment (hfiguiere) [inline org.godotengine.Godot3.yaml:93]:
```suggestion
      - install -Dm644 misc/dist/linux/org.godotengine.Godot.xml /app/share/mime/packages/$FLATPAK_ID.xml
```

File: org.godotengine.Godot3.yaml (hunk 6 above)
Comment (hfiguiere) [reply org.godotengine.Godot3.yaml:83]: As I said, you should use upstream's file and patch it. Not a local copy.

**Comment by Zishan-Rahman** [comment]
I'm going to try building it _without_ removing the duplicate appdata.xml file first, then I'll try and figure out how I can use the upstream's appdata.xml file instead.

**Comment by Zishan-Rahman** [comment]
bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
> Build [28910 successful](https://buildbot.flathub.org/#/builders/6/builds/28910) To test this build, install it from the testing repository:
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/11575/org.godotengine.Godot3.flatpakref
> ```

Wow, that _worked_?

The Flatpak works fine, but we still have that duplicate appdata.xml to take care of. I just pushed a couple of commits which should hopefully negate the need for a duplicate appdata.xml. If the next build works (with these new changes), we can remove the existing appdata.xml file. After that, we _should_ be able to push this new Godot 3 Flatpak to Flathub, provided everything works, of course.

**Comment by Zishan-Rahman** [comment]
For hopefully the last time, bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
```
Applying patch update-appdata.patch
FB: Running: patch -p4 -i /srv/buildbot/worker/build-x86_64-1/build/update-appdata.patch
can't find file to patch at input line 4
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/misc/dist/linux/org.godotengine.Godot.appdata.xml b/misc/dist/linux/org.godotengine.Godot.appdata.xml
|--- a/misc/dist/linux/org.godotengine.Godot.appdata.xml	2023-03-06 14:13:41.000000000 +0000
|+++ b/misc/dist/linux/org.godotengine.Godot.appdata.xml	2023-03-16 22:20:25.946962567 +0000
--------------------------
```

Darn.

**Comment by Zishan-Rahman** [comment]
Maybe _that_ should do it?

Don't fail me now bot, build org.godotengine.Godot3

File: org.godotengine.Godot3.yaml [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      - type: script
+        dest-filename: godot.sh
+        commands:
+          - export APPDATA="$XDG_DATA_HOME"
+          - export PATH="/app/jre/bin:$PATH"
+          - /app/bin/godot-bin "$@"
+
+      - type: patch
+        path: update-appdata.patch
+        strip-components: 0
```
Comment (hfiguiere) [inline org.godotengine.Godot3.yaml:84]:
```suggestion
```

**Comment by Zishan-Rahman** [comment]
Eh, I'll send a _third_ build to the Flathub build server, why not?

Hey bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
Can anyone with Flathub commit permissions or admin status please [stop this build](https://buildbot.flathub.org/#/builders/6/builds/28921) so our new build in the queue can hopefully get processed as soon as possible? Unfortunately I can't stop it myself.

File: org.godotengine.Godot3.yaml [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        path: update-appdata.patch
+
+    build-commands:
+      - python3 /app/bin/scons $SCONS_FLAGS $SCONS_FLAGS_EXTRA tools=yes target=release_debug -j "$FLATPAK_BUILDER_N_JOBS"
+      - install -D -m755 bin/godot.x11.opt.tools.* /app/bin/godot-bin
+      - install -D -m755 godot.sh /app/bin/godot
+      - desktop-file-edit --set-icon=${FLATPAK_ID} misc/dist/linux/org.godotengine.Godot.desktop
+      - install -Dm644 misc/dist/linux/org.godotengine.Godot.desktop /app/share/applications/$FLATPAK_ID.desktop
+      - install -Dm644 misc/dist/linux/org.godotengine.Godot.xml /app/share/mime/packages/$FLATPAK_ID.xml
+      - install -Dm644 $FLATPAK_ID.appdata.xml /app/share/appdata/$FLATPAK_ID.appdata.xml
```
Comment (hfiguiere) [inline org.godotengine.Godot3.yaml:92]:
```suggestion
      - install -Dm644 misc/dist/linux/org.godotengine.Godot.appdata.xml /app/share/appdata/$FLATPAK_ID.appdata.xml
```

**Comment by Zishan-Rahman** [comment]
Come on now, bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
I just opened and ran a 3D project I had on my computer and the Flatpak works perfectly fine. Looks like we can delete the redundant appdata.xml (it's no longer defined in the YAML anyway so it'll make no difference). For good measure, I will build it once more and, if all goes to plan, we _should_ be able to get this Godot 3 Flatpak onto Flathub today!

**Comment by Zishan-Rahman** [comment]
For the last time (and please let it be the last time), bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
Eh, one more build, for good measure. I know I only updated the README here but still.

bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
This_should_ work and when it does, we can finally get this up on Flathub! :relieved:

**Comment by Zishan-Rahman** [comment]
I just realised that [due to this change](https://github.com/flathub/flathub/pull/3988/commits/e925d000ccddc76398e81b602316787b2ee3d7af), we can no longer set up external code editor use as described in the README.md. However, reintroducing the line that was deleted in that change would result in a "finish-args-flatpak-spawn-access" error in flatpak-builder-lint. This is the only really outstanding issue here, so I'm not sure how to fix it again, or if I should just include a note in the README and appdata.xml files. This is the only _really_ outstanding issue here, otherwise we're good to go.

**Comment by hfiguiere** [comment]
> Weirdly enough, when I add the spawn access back in through either Flatseal or the command line, it works fine, but for some reason flatpak-builder-lint doesn't like it when it's included in the manifest itself.

because it's a sandbox hole and an exception must be applied for as to prevent to have them randomly added for no reason.

**Comment by Zishan-Rahman** [comment]
> > Weirdly enough, when I add the spawn access back in through either Flatseal or the command line, it works fine, but for some reason flatpak-builder-lint doesn't like it when it's included in the manifest itself.
> 
> because it's a sandbox hole and an exception must be applied for as to prevent to have them randomly added for no reason.

Thanks @hfiguiere. How do I apply this exception and still allow flatpak-spawn access inside the YAML (or did I misread your reply)? Godot is going to need access to flatpak-spawn if external editor support is to work.

**Comment by Zishan-Rahman** [comment]
>  ~~How do I apply this exception and still allow flatpak-spawn access inside the YAML (or did I misread your reply)?~~

I misread your reply, didn't I? I probably misinterpreted it because of the word _exception_ when flatpak-builder-lint is throwing an _error_. Of course wanton flatpak-spawn access is going to be a security issue, but this is a Flatpak that needs it (for external editor support).

I'm not knowledgeable as to how to add these justifications to our YAML or anywhere else in our repository (I'm still new to this and still learning). Still, there's _got_ to be a way for us to get flatpak-spawn access back in there out-of-the-box, especially since, as I said, the existing Godot 4 Flatpak has the needed support and does not throw this error.

**Comment by Zishan-Rahman** [comment]
I added notes (well, warnings) about the state of editor support for now, until we can either (1) get it fixed before merging to Flathub or (2) leave it to the side (hopefully temporarily).

Alright bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
@hfiguiere @Calinou @barthalion @TingPing I apologise in advance for the surprise pings, but I'm having an issue that needs resolving.

This Godot 3 Flatpak is almost ready... except for one significant regression. It's all explained in the README.md and the past commit (as well as the recent comments here), but [I'll quote a prior comment of mine](https://github.com/flathub/flathub/pull/3988#issuecomment-1473664903) for context: 

> I just realised that [due to this change](https://github.com/flathub/flathub/pull/3988/commits/e925d000ccddc76398e81b602316787b2ee3d7af), we can no longer set up external code editor use as described in the README.md. However, reintroducing the line that was deleted in that change [would result in a "finish-args-flatpak-spawn-access" error in flatpak-builder-lint](https://github.com/flathub/flatpak-builder-lint/blob/e5105120d28614f385769d9f2aa31029f5e89e9f/flatpak_builder_lint/checks/finish_args.py#L86), and that was why it was removed in the first place. I'm not sure how to fix it again, or if I should just include a note in the README and appdata.xml files. This is the only _really_ outstanding issue here, otherwise we're good to go.

As @hfiguiere and I have already mentioned here, the reason for the "finish-args-flatpak-spawn-error" is that allowing random Flatpaks to wantonly access flatpak-spawn pokes through the sandbox and poses a potential security risk. But this Flatpak _needs_ flatpak-spawn access for external editor support to work. And what's weird is that the Godot 4 Flatpak doesn't seem to be having this problem. [The offending line is still in there and not a flatpak-builder-lint error in sight](https://github.com/flathub/org.godotengine.Godot/blob/f74ec64e2933a5304dcc0b4228985f5ec8ed96f2/org.godotengine.Godot.yaml#L49), so I don't know where else this problem could be arising. Nor do I know why it's happening _here_ but not _there_.

[In my most recent commit](https://github.com/flathub/flathub/pull/3988/commits/3aa82b01fd0fe17fb77d02bbbeb11ec273d6f911), I made changes to the README.md and my appdata.xml patch to make future users aware of this issue, just in case we leave it for now and try and fix it after it's published, but this is not ideal, and I'd much rather we fix this problem before the Godot 3 Flatpak goes live. If we _do_ get it fixed before this goes live, I can revert that most recent commit. If any one of you has any kind of solution, or has advice on how to fix this issue, please do post here. I will be really grateful for any help you can give. Anyone else I _didn't_ ping who can help me I'll be thankful for as well.

**Comment by TingPing** [comment]
@Zishan-Rahman It has to be granted an exception here: https://github.com/flathub/flatpak-builder-lint/blob/fc76448d5539fb6a42f1a0b6f4a8982beef17c9e/flatpak_builder_lint/staticfiles/exceptions.json#L1185-L1187

**Comment by Zishan-Rahman** [comment]
> @Zishan-Rahman It has to be granted an exception here: https://github.com/flathub/flatpak-builder-lint/blob/fc76448d5539fb6a42f1a0b6f4a8982beef17c9e/flatpak_builder_lint/staticfiles/exceptions.json#L1185-L1187
> 
> You can open a PR and/or barthalion will look at it.

Darn, how did I not know about this earlier? :man_facepalming: 
I'll open a PR there tonight.

**Comment by Zishan-Rahman** [comment]
> > Weirdly enough, when I add the spawn access back in through either Flatseal or the command line, it works fine, but for some reason flatpak-builder-lint doesn't like it when it's included in the manifest itself.
> 
> because it's a sandbox hole and an exception must be applied for as to prevent to have them randomly added for no reason.

So _that_'s what hfiguere meant by _exception_! :man_facepalming:

**Comment by hfiguiere** [comment]
sorry I was meaning to clarify that.

**Comment by barthalion** [comment]
bot, build org.godotengine.Godot3

@Calinou, can you review if it makes sense to you?

**Comment by Zishan-Rahman** [comment]
> bot, build org.godotengine.Godot3
> 
> @Calinou, can you review if it makes sense to you?

@barthalion I just pushed some commits over that I made a week ago; I was meant to push them _after_ the merge, but I forgot.

The ongoing build won't work with external editor permissions out of the box, because the "new" commits had to reintroduce these permission. We'll need to build it again after that one is done.

**Comment by Zishan-Rahman** [comment]
> Queued test build for org.godotengine.Godot3.

Oh.

File: org.godotengine.Godot3.yaml [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  - org.freedesktop.Sdk.Extension.openjdk11
+command: godot
+
+build-options:
+  arch:
+    x86_64:
+      env:
+        # Only enable link-time optimization when targeting x86_64
+        # (causes issues on other architectures)
+        SCONS_FLAGS_EXTRA: use_lto=yes
```
Comment (Calinou) [inline org.godotengine.Godot3.yaml:15]: Use up-to-date SCons option for enabling LTO:

```suggestion
        SCONS_FLAGS_EXTRA: lto=full
```

**Review by Calinou — COMMENTED** [review]
Should be good to merge after applying the suggestion.

**Comment by Zishan-Rahman** [comment]
I guess we're doing a _third_ test build? :thinking:

**Comment by Zishan-Rahman** [comment]
Actually the second one's still in the queue so we can just overwrite it.

For (hopefully) the last time, bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
Build [118237](https://buildbot.flathub.org/#/buildrequests/118237) is still waiting.

**Comment by Zishan-Rahman** [comment]
Opened a Godot project, then tried opening a GDScript file in VSCodium _through_ Godot, and it worked! Look like we can _finally_ merge this now! :relieved:

**Comment by barthalion** [comment]
The desktop file should also mention it's Godot 3.x, currently it's the vanilla "Godot Engine" name.

**Comment by Zishan-Rahman** [comment]
> The desktop file should also mention it's Godot 3.x, currently it's the vanilla "Godot Engine" name.

Ah. Working on it.

**Comment by Zishan-Rahman** [comment]
Done.

bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
[I figured _this_ would be better](https://github.com/flathub/flathub/pull/3988/commits/7e0db922e198f4851ec30a2c91b0ab2f7476cf2b).

OK now, bot, build org.godotengine.Godot3

**Comment by Zishan-Rahman** [comment]
> The desktop file should also mention it's Godot 3.x, currently it's the vanilla "Godot Engine" name.

@barthalion Just got this sorted out. If nothing else needs fixing it looks like we can finally merge this into Flathub! :relieved:

**Pull request closed without merging**